src/Routing.jl

```diff
@@ -0,0 +1,220 @@
+# general methods for routing rivers
+export compute_streamflow!,
+    update_state_from_hillslope!, update_state_from_channel!
+
+using CSV, DataFrames, Dates, DSP, SpecialFunctions
+
+# methods
+function compute_streamflow!(
+    river_state::RS,
+    river_model::HCM,
+    environment::E,
+    start_date::Date,
+    end_date::Date,
+) where {RS <: RiverState, HCM <: HillslopeChannelRiverModel, E <: Environment}
+
+    update_state_from_hillslope!(
+        river_state,
+        river_model.hillslope_model,
+        environment,
+        start_date,
+        end_date,
+    )
+    update_state_from_channel!(
+        river_state,
+        river_model.channel_model,
+        environment,
+        start_date,
+        end_date,
+    )
+    # return calculate_streamflow(river_state)
+end
+
+
+# Auxiliary function
+## function for reading basins from txt file into Vector{Int}
+function read_routing_lv_basins(routing_lv_file::String)
+    routing_lv_basins = Int64[]
+    file = open(joinpath(routing_lv_file))
+    for line in eachline(file)
+        push!(routing_lv_basins, parse(Int64, line))
+    end
+    close(file)
+
+    return routing_lv_basins
+end
+
+# only for one routing lv
+function update_state_from_hillslope!(
+    river_state::RS,
+    hillslope_model::HM,
+    env::E,
+    start_date::Date,
+    end_date::Date,
+) where {RS <: RiverState, HM <: AbstractHillslopeModel, E <: Environment}
+    println("Starting hillslope update")
+
+    # Constants
+    day_to_s = 86400
+    km²_to_m² = 1000000
+
+    timeseries_dir = env.dyn_env.timeseries_dir
+    output_dir = env.dyn_env.output_dir
+    dates = collect(start_date:Day(1):end_date)
+
+    routing_lv_basins_dir = env.static_env.routing_lv_basins_dir
+    attributes_df = CSV.read(
+        joinpath(env.static_env.attributes_dir, "attributes.csv"),
+        DataFrame,
+    )
+
+    a, θ = hillslope_model.shape, hillslope_model.timescale
+    t_max = hillslope_model.t_max
+
+    distribution = [
+        (t^(a - 1) * exp(-t / θ)) / (θ^a * SpecialFunctions.gamma(a)) for
+        t in 0:(t_max - 1)
+    ]
+
+    all_basin_ids = read_routing_lv_basins(
+        joinpath(routing_lv_basins_dir, "all_basin_ids.txt"),
+    )
+    # make new txt file 
+    for basin_id in all_basin_ids # routing_lv_basins
+        timeseries_df =
+            CSV.read(joinpath(timeseries_dir, "basin_$basin_id.csv"), DataFrame)
+        filtered_df =
+            filter(row -> start_date <= row[:date] <= end_date, timeseries_df)
+        basin_area =
+            attributes_df[attributes_df.HYBAS_ID .== basin_id, :area][1]
+        runoff =
+            (filtered_df[:, :sro_sum] .+ filtered_df[:, :ssro_sum]) .*
+            basin_area ./ day_to_s .* km²_to_m²
+
+        streamflow = DSP.conv(runoff, distribution)[1:size(runoff)[1], :][:]
+
+        output_df = DataFrame(date = dates, streamflow = streamflow)
+
+        CSV.write(
+            joinpath(output_dir, "hillslope_basin_$basin_id.csv"),
```

Comment:
This seems correct, but if we are at a loss as to where the discrepancy comes in, we could double check one basin in this file and compute the convolution by hand.